feat: update password service

oopshell opened the pull request:
#### Added resolver of changeing password:
- get ID of the current logged in user;
- verify current password;
- replace old password with new one
- add password pattern validation pipe

#### Added unit test:
- add test for changePassword service
- add test for password pattern validation pipe

#### Resolve CP-34

![Screenshot 2024-04-17 at 03 30 28](https://github.com/BeeQuantAI/platform_api/assets/48723564/33cd552d-6046-4780-8484-500f082e9924)
![Screenshot 2024-04-17 at 03 31 01](https://github.com/BeeQuantAI/platform_api/assets/48723564/5cedc488-3560-4688-8620-0b8f4dc128cc)
![Screenshot 2024-05-29 at 17 46 10](https://github.com/BeeQuantAI/platform_api/assets/48723564/d4f11e07-16f4-4abb-9c1a-7594b5c6cab4)

Chris-Wang commented on src/modules/auth/auth.resolver.ts:
```diff
@@ -80,4 +83,40 @@ export class AuthResolver {
+  @Mutation(() => Result, { description: 'Change password' })
+  @UseGuards(GqlAuthGuard)
+  async changePassword(
+    @Context() cxt: any,
```
avoid `any`, add context type

Chris-Wang commented on src/modules/auth/auth.resolver.ts:
```diff
@@ -80,4 +83,40 @@ export class AuthResolver {
+    }
+    const isPasswordValid = await bcrypt.compare(oldPassword, user.password);
+    if (isPasswordValid) {
+      const hashedPassword = await bcrypt.hash(newPassword, 10);
```
consider add new password validation

Chris-Wang commented on src/modules/auth/auth.resolver.ts:
```diff
@@ -1,11 +1,14 @@
 import {
   ACCOUNT_EXIST,
   ACCOUNT_NOT_EXIST,
+  UPDATE_ERROR,
```
to be more specific, I think we need to add `UPDATE_PASSWORD_ERROR`

Chris-Wang commented on src/modules/auth/auth.resolver.ts:
```diff
@@ -80,4 +83,40 @@ export class AuthResolver {
+    @Context() cxt: any,
+    @Args('oldPassword') oldPassword: string,
+    @Args('newPassword') newPassword: string,
+  ): Promise<Result> {
```
let's add try-catch block in the resolver to catch any unexpected error, and return the error text in response

Chris-Wang reviewed (changes requested):
Please add unit test

oopshell commented:
> Please add unit test

Hi @Chris-Wang , I used **reset --hard** to discard some clumsy commits, and now one of your review comments becomes unresolvable. Could you kindly re-review the PR? I apologise for the trouble and appreciate your help. Thank you!

Chris-Wang commented on src/modules/auth/auth.resolver.ts:
```diff
@@ -26,4 +27,14 @@ export class AuthResolver {
+  @Mutation(() => Result, { description: 'Change password' })
+  @UseGuards(GqlAuthGuard)
+  async changePassword(
+    @Context() cxt: { req: Partial<Request> & { user: {id: string} } },
```
`{ user: { id: string } }` leave space after or before braces

Chris-Wang commented on src/modules/auth/auth.service.ts:
```diff
@@ -69,4 +71,46 @@ export class AuthService {
   }
+
+  async changePassword(
+    cxt: { req: Partial<Request> & { user: {id: string} } },
```
`{ user: { id: string } }` leave space after or before braces

Chris-Wang commented on src/modules/auth/pipe/password-validation.pipe.ts:
```diff
@@ -0,0 +1,21 @@
+    'string.pattern.base': 'password must contain 8 to 32 characters, including letter, number and special character.'
+  });
+
+  transform(value: any): string {
```
consider to use custom exception to give non-error response, 
now we get:
<img width="716" alt="image" src="https://github.com/BeeQuantAI/platform_api/assets/1610136/910d7532-e403-499d-8dec-265e1726b812">
here's like what we want:
<img width="509" alt="image" src="https://github.com/BeeQuantAI/platform_api/assets/1610136/4ac1724d-e676-4335-843f-15a8494ace1f">

oopshell replied:
![Screenshot 2024-05-29 at 17 46 10](https://github.com/BeeQuantAI/platform_api/assets/48723564/fea9d00a-4595-469f-b575-dce4f7d8a9a7)

oopshell committed:
feat: update password service

Added the resolver and service of changePassword:
- get ID of the current logged in user;
- verify current password;
- replace old password with new one
- add password pattern validation pipe

Added unit test:
- add test for changePassword service
- add test for password pattern validation pipe

Resolve CP-34

Chris-Wang reviewed (approved):
LGTM
